contracts/utils/structs/MerkleTree.sol

```diff
@@ -0,0 +1,160 @@
+     * - Depth `depth`
+     * - All leaves are initialize to `zero`
+     * - Hashing function for a pair of leaves is fnHash.
+     *
+     * If the MerkleTree was already setup and used, calling that function again will reset it to a blank state.
+     */
+    function setup(
+        Bytes32MerkleTree storage self,
+        uint256 depth,
+        bytes32 zero,
+        function(bytes32, bytes32) view returns (bytes32) fnHash
+    ) internal {
+        if (depth > MAX_DEPTH) {
+            revert MerkleTreeInvalidDepth(depth, MAX_DEPTH);
+        }
+
+        // Store depth in the dynamic array
+        Arrays.unsafeSetLength(self.sides, depth);
+        Arrays.unsafeSetLength(self.zeros, depth);
+
+        // Build the different hashes in a zero-filled complete tree
+        bytes32 currentZero = zero;
+        for (uint32 i = 0; i < depth; ++i) {
+            Arrays.unsafeAccess(self.zeros, i).value = currentZero;
+            currentZero = fnHash(currentZero, currentZero);
+        }
+        // Set the first root
+        self.root = currentZero;
+        self.nextLeafIndex = 0;
+        self.fnHash = fnHash;
+    }
+
+    /**
+     * @dev Insert a new leaf in the tree, and compute the new root.
+     */
+    function insert(Bytes32MerkleTree storage self, bytes32 leaf) internal returns (uint256) {
+        // Cache read
+        uint256 depth = self.zeros.length;
+        function(bytes32, bytes32) view returns (bytes32) fnHash = self.fnHash;
+
+        // Get leaf index
+        uint256 leafIndex = self.nextLeafIndex++;
+
+        // Check if tree is full.
+        if (leafIndex >= 1 << depth) {
+            Panic.panic(Panic.RESOURCE_ERROR);
+        }
+
+        // Rebuild branch from leaf to root
+        uint256 currentIndex = leafIndex;
+        bytes32 currentLevelHash = leaf;
+        for (uint32 i = 0; i < depth; i++) {
+            // Reaching the parent node, is currentLevelHash the left child?
+            bool isLeft = currentIndex % 2 == 0;
+
+            // If so, next time we will come from the right, so we need to save it
+            if (isLeft) {
+                Arrays.unsafeAccess(self.sides, i).value = currentLevelHash;
+            }
+
+            // Compute the node hash by hashing the current hash with either:
+            // - the last value for this level
+            // - the zero for this level
+            currentLevelHash = fnHash(
+                isLeft ? currentLevelHash : Arrays.unsafeAccess(self.sides, i).value,
+                isLeft ? Arrays.unsafeAccess(self.zeros, i).value : currentLevelHash
+            );
+
+            // Update node index
+            currentIndex >>= 1;
+        }
+
+        // Record new root
+        self.root = currentLevelHash;
+
+        return leafIndex;
+    }
+
+    /**
+     * @dev Tree's depth (set at initialization)
+     */
+    function getDepth(Bytes32MerkleTree storage self) internal view returns (uint256) {
```

Comment:
I'd rename this as just `depth`:

```suggestion
    function depth(Bytes32MerkleTree storage self) internal view returns (uint256) {
```

What do you think?

Reply:
that creates naming conflicts, doesn't it?